server/src/main/java/org/opensearch/index/shard/IndexShard.java

```diff
@@ -4833,39 +4833,96 @@ private String copySegmentFiles(
         Map<String, RemoteSegmentStoreDirectory.UploadedSegmentMetadata> uploadedSegments,
         boolean overrideLocal
     ) throws IOException {
-        List<String> downloadedSegments = new ArrayList<>();
-        List<String> skippedSegments = new ArrayList<>();
+        Set<String> toDownloadSegments = new HashSet<>();
+        Set<String> skippedSegments = new HashSet<>();
         String segmentNFile = null;
+
         try {
-            Set<String> localSegmentFiles = Sets.newHashSet(storeDirectory.listAll());
             if (overrideLocal) {
-                for (String file : localSegmentFiles) {
+                for (String file : storeDirectory.listAll()) {
                     storeDirectory.deleteFile(file);
                 }
             }
+
             for (String file : uploadedSegments.keySet()) {
                 long checksum = Long.parseLong(uploadedSegments.get(file).getChecksum());
                 if (overrideLocal || localDirectoryContains(storeDirectory, file, checksum) == false) {
-                    storeDirectory.copyFrom(sourceRemoteDirectory, file, file, IOContext.DEFAULT);
-                    downloadedSegments.add(file);
+                    toDownloadSegments.add(file);
                 } else {
                     skippedSegments.add(file);
                 }
-                if (targetRemoteDirectory != null) {
-                    targetRemoteDirectory.copyFrom(storeDirectory, file, file, IOContext.DEFAULT);
-                }
+
                 if (file.startsWith(IndexFileNames.SEGMENTS)) {
                     assert segmentNFile == null : "There should be only one SegmentInfosSnapshot file";
                     segmentNFile = file;
                 }
             }
+
+            if (toDownloadSegments.isEmpty() == false) {
+                try {
+                    final CountDownLatch completionLatch = new CountDownLatch(toDownloadSegments.size());
```

Comment:
We don't need a CountDownLatch of size equal to number of files. We just need it of size 1 just to notify that download of all files have completed. Since, latch countdown is an atomic operation, it will block other callbacks. You can use GroupedActionListener to handle multiple callbacks. You can refer RemoteStoreRefreshListener#uploadNewSegments .

Reply:
GroupedActionListener encapsulates a CountDownLatch, it will just ensure a callback is done once all operations are complete. However it will allow other operations to proceed. Fundamentally the same, however I am more concerned about which threadpool does the await.

Reply:
One minor difference is that there will be n atomic countdowns vs 1. Await will happen in the current thread itself, whichever thread is invoking sync segments.

Reply:
Yes, forgot that it's not aggregating listeners and is just a wrapper for n sized latch+listener.

Reply:
Updated the code to use `GroupedActionListener` instead. Thanks for the inputs.